### PR TITLE
Staging extensions

### PR DESCRIPTION
**Pull request by habib-sadullaev**

### CODE REVIEW

File: src/TypeShape/StagingExtensions.fs [hunk 1]
```diff
@@ -135,14 +128,13 @@ module Expr =
...
-    /// `(fun x y z .. -> M[x,y,z,..]) a b c => M[a,b,c,..]`,
-    /// where a,b,c are variables or constants
+    /// `(fun x y z .. -> M[x,y,z,..]) a b c` => `M[a,b,c,..]`
     let unlambda (expr : Expr<'T>) =
         let (|AppLambdas|_|) (e : Expr) =
             // traverses the "App(App(App ... " part of the expression
             let rec gatherApps args e =
                 match e with
-                | Application(lhs, ((Var _ | Value _) as rhs)) -> gatherApps (rhs :: args) lhs
+                | Application(lhs, rhs) -> gatherApps (rhs :: args) lhs
```
Comment (eiriktsarpalis) [inline src/TypeShape/StagingExtensions.fs:137]: Why the change? If the right-hand-side expression is impure then this is not a valid reduction.

Comment (habib-sadullaev) [reply src/TypeShape/StagingExtensions.fs:137]: Yes, agreed. But this reduction is useful in some scenario.
For example:
```fsharp
type Foo = { Value : int; Name: string }
let x  = Var("x"; typeof<Foo>) |> Expr.Var |> Expr.Cast<Foo>

let eq = <@ (=) @>
let const = <@ 1 @>
let expr1 = <@ (%eq) (%x).Value %c  @>

let startsWith = <@ fun (x: string) (y: string) -> x.StartsWith y @>
let substr  = <@ "abc" @>
let expr2 = <@ (%startsWith) (%x).Name %substr @> 
```
Or in my code: [declaration](https://github.com/habib-sadullaev/Equal/blob/81a991961115bfe3caf7bfa4614458a50251845c/src/Equal/Operators.fs#L16) and [usage](https://github.com/habib-sadullaev/Equal/blob/master/src/Equal/Expression.fs#L96)

Comment (eiriktsarpalis) [reply src/TypeShape/StagingExtensions.fs:137]: It violates the evaluation semantics of F# and would be a breaking change. I therefore would not be able to accept this even if it could be meaningful in certain scenaria.

Comment (habib-sadullaev) [reply src/TypeShape/StagingExtensions.fs:137]: how about adding a new function, say ``unlambdaForced`` instead of changing ``unlambda``?

Comment (eiriktsarpalis) [reply src/TypeShape/StagingExtensions.fs:137]: It seems like a fairly straightforward variation to include as a helper function in your code.

Comment (habib-sadullaev) [reply src/TypeShape/StagingExtensions.fs:137]: yes, I already did this). what about my other changes, if I revert this one?

File: src/TypeShape/StagingExtensions.fs [hunk 2]
```diff
@@ -117,11 +112,9 @@ module Expr =
     /// Expands a collection of state-updating staged computations
     /// into a expression tree
     let update (varName : string, init : Expr<'T>) (comps : (Expr<'T> -> Expr<'T>) []) : Expr<'T> =
-        if comps.Length = 0 then init else
-        bindMutable (varName, init)
-            (fun getter setter -> 
-                let unfolded = comps |> Array.map (fun c -> setter (c getter)) |> seq
-                <@ %unfolded ; %getter @>)
+        let inputs = Array.map lam comps
```
Comment (eiriktsarpalis) [inline src/TypeShape/StagingExtensions.fs:115]: Please revert this change, lifting to `lam` introduces unecessary lambda abstractions in the expression tree.

File: src/TypeShape/StagingExtensions.fs [hunk 3]
```diff
@@ -182,7 +174,7 @@ module Expr =
     let unlet (expr : Expr<'T>) : Expr<'T> =
         let rec aux e =
             match e with
-            | Let(x, (Var _ | Value _ as e), body) when not x.IsMutable ->
+            | Let(x, (Var _ | Value _ as e), body) when not x.IsMutable || x.Type.IsValueType && x.Name = "copyOfStruct" ->
```
Comment (eiriktsarpalis) [inline src/TypeShape/StagingExtensions.fs:178]: Please revert this change.

File: src/TypeShape/StagingExtensions.fs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/TypeShape/StagingExtensions.fs]: I think replacing the `Array.fold[Back]` calls with `Array.reduce[Back]` are good improvements.

File: src/TypeShape/StagingExtensions.fs (hunk 3 above)
Comment (habib-sadullaev) [reply src/TypeShape/StagingExtensions.fs]: done

File: src/TypeShape/StagingExtensions.fs (hunk 2 above)
Comment (habib-sadullaev) [reply src/TypeShape/StagingExtensions.fs]: done

**Comment by eiriktsarpalis** [comment]
Thanks!
